Third and forth tasks

DariaMaslova opened the pull request.

Commit:
task 3: Main navigation implemented

- Navigation through the app by AppNavigator
- Home page navigation (drawer) by MenuNavigator

Commit:
task 4: loads plant from network

- implemented Catalogue Page

Commit:
task 3: returns result from screen

- Returns and applies result from filter page

GamaGong commented on lib/api/model/species.dart:
```diff
@@ -0,0 +1,94 @@
+      id: json["id"],
+      commonName: json["common_name"],
+      slug: json["slug"],
+      scientificName: json["scientific_name"],
```
Можно кидать эксепшн, если значение всё же `null`. Не критично

DariaMaslova replied:
Не совсем поняла, к какой строке относится комментарий

GamaGong replied:
ты рассчитываешь на не null значение в этом поле, json может его спарсить в null, насколько я понимаю. Поэтому можно было бы кинуть ошибку, если он это сделает.
```
final String scientificName; // Non null
```

GamaGong commented on lib/core/extensions/cap_extensions.dart:
```diff
@@ -0,0 +1,16 @@
+extension CapExtension on String {
+  String get inCaps => '${this[0].toUpperCase()}${this.substring(1)}';
```
тут не нужна аналогичная проверка как в `get capitalize` ?

DariaMaslova replied:
Удалила функцию, т.к. дублирует capitalize

GamaGong commented on lib/main.dart:
```diff
@@ -1,22 +1,93 @@
+
+  /// The app state contains a key for the app level navigator in the
+  /// [MaterialApp] widget which can be used to push, pop and replace routes
+  final navKey = GlobalKey<NavigatorState>();
```
насколько я понимаю, этот код нужен чтобы получить доступ к рутовому `Navigatior`. Если так, то можно использовать `Navigator.of(context, rootNavigator: true);`

DariaMaslova replied:
Спасибо. Заменила

GamaGong commented on lib/pages/catalogue/catalogue_page.dart:
```diff
@@ -0,0 +1,190 @@
+                  child: _Catalogue(
+                    plants: state.plants ??
+                        List<CataloguePlant>(),
+                    onLoadMoreListener: () {
```
для однострочных лямбд есть синтаксис `() => foo()`
Тут же вообще можно использовать ссылку на функцию `_viewModel.loadMore`

DariaMaslova replied:
Заменила

GamaGong commented on lib/pages/catalogue/catalogue_viewmodel.dart:
```diff
@@ -0,0 +1,161 @@
+
+  subscribeFilter(FilterController filterController) {
+    if (_state == null) {
+      return; // not initialized
```
по идее это исключительный случай, можно и ошибку кинуть

DariaMaslova replied:
Добавила IllegalStateException

Commit:
fixes comments

- uses one line lambdas
- adds IllegalStateException
- removes inCaps function
- uses Navigator.of(context, rootNavigator: true)
